web3/contract.py

```diff
@@ -1328,6 +1337,45 @@ def factory(cls, class_name, **kwargs):
         return PropertyCheckingFactory(class_name, (cls,), kwargs)
 
 
+class ContractCaller:
+    def __init__(self, abi, web3, address, *args, transaction_dict=None, **kwargs):
+        self.web3 = web3
+        self.abi = abi
+        self.address = address
+
+        if transaction_dict is None:
+            transaction_dict = {}
+
+        if self.abi:
+            self._functions = filter_by_type('function', self.abi)
+            for func in self._functions:
+                fn = ContractFunction.factory(
+                    func['name'],
+                    web3=self.web3,
+                    contract_abi=self.abi,
+                    address=self.address,
+                    function_identifier=func['name'])
+
+                caller_method = partial(self.call_function, fn, transaction_dict=transaction_dict)
+                setattr(self, func['name'], caller_method)
+        else:
+            raise NoABIFunctionsFound(
```

Comment:
I think this is the culprit.  This needs to be lazy.  See how `ContractFunctions` implements a `__getattr__` which gives it a chance to raise the exception anytime a user tries to access a function on the object.

Reply:
Aha! Thanks. Yeah, this just started causing problems.